source/extensions/filters/network/zookeeper_proxy/zookeeper_decoder.cc

```diff
@@ -0,0 +1,344 @@
+  std::string path = BufferHelper::peekString(data, offset);
+
+  // Skip data.
+  int32_t datalen = BufferHelper::peekInt32(data, offset);
+  offset += datalen;
+  skipAcls(data, offset);
+  int32_t flags = BufferHelper::peekInt32(data, offset);
+  const bool ephemeral = (flags & 0x1) == 1;
+  const bool sequence = (flags & 0x2) == 2;
+
+  callbacks_.onCreateRequest(path, ephemeral, sequence, two);
+}
+
+void DecoderImpl::parseSetRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  checkLength(len, 20);
+
+  std::string path = BufferHelper::peekString(data, offset);
+  // Skip data.
+  int32_t datalen = BufferHelper::peekInt32(data, offset);
+  offset += datalen;
+  // Ignore version.
+  BufferHelper::peekInt32(data, offset);
+
+  callbacks_.onSetRequest(path);
+}
+
+void DecoderImpl::parseGetChildrenRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len,
+                                          const bool two) {
+  checkLength(len, 14);
+
+  std::string path = BufferHelper::peekString(data, offset);
+  bool watch = BufferHelper::peekBool(data, offset);
+
+  callbacks_.onGetChildrenRequest(path, watch, two);
+}
+
+void DecoderImpl::parseDeleteRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  checkLength(len, 16);
+
+  std::string path = BufferHelper::peekString(data, offset);
+  int32_t version = BufferHelper::peekInt32(data, offset);
+
+  callbacks_.onDeleteRequest(path, version);
+}
+
+void DecoderImpl::parseExistsRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  checkLength(len, 13);
+
+  std::string path = BufferHelper::peekString(data, offset);
+  bool watch = BufferHelper::peekBool(data, offset);
+
+  callbacks_.onExistsRequest(path, watch);
+}
+
+void DecoderImpl::parseGetAclRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  checkLength(len, 8);
+
+  std::string path = BufferHelper::peekString(data, offset);
+
+  callbacks_.onGetAclRequest(path);
+}
+
+void DecoderImpl::parseSetAclRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  checkLength(len, 8);
+
+  std::string path = BufferHelper::peekString(data, offset);
+  skipAcls(data, offset);
+  int32_t version = BufferHelper::peekInt32(data, offset);
+
+  callbacks_.onSetAclRequest(path, version);
+}
+
+void DecoderImpl::parseSyncRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  checkLength(len, 8);
+
+  std::string path = BufferHelper::peekString(data, offset);
+
+  callbacks_.onSyncRequest(path);
+}
+
+void DecoderImpl::parseCheckRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  checkLength(len, 8);
+
+  std::string path = BufferHelper::peekString(data, offset);
+  int32_t version = BufferHelper::peekInt32(data, offset);
+
+  callbacks_.onCheckRequest(path, version);
+}
+
+void DecoderImpl::parseMultiRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  // Treat empty transactions as a decoding error, there should be at least 1 header.
```

Comment:
Hard for me as reader to understand how "there should be at least 1 header" relates to checking for a length of at least 17